mapper/collectd_mapper/src/batcher.rs

```diff
@@ -1,169 +1,171 @@
+            });
+
         Self {
             sender,
             mqtt_client,
             source_topic_filter,
-            batching_window,
+            batcher,
             clock,
         }
     }
 
-    pub async fn run(&self) -> Result<(), DeviceMonitorError> {
+    pub async fn run(&mut self) -> Result<(), DeviceMonitorError> {
         let mut messages = self
             .mqtt_client
             .subscribe(self.source_topic_filter.clone())
             .await?;
 
+        let mut outputs = Vec::new();
+
         loop {
-            match self.receive_message(messages.as_mut()).await {
-                Some((message, timestamp)) => {
-                    // Build a message batch until the batching window times out and return the batch
-                    let message_batch_result = self
-                        .build_message_batch_with_timeout(message, timestamp, messages.as_mut())
-                        .await;
-
-                    match message_batch_result {
-                        Ok(message_batch) => {
-                            // Send the current batch to the batch processor
-                            let _ = self.sender.send(message_batch).map_err(|err| {
+            let next_notify_in = self.process_outputs(&mut outputs);
+            self.process_io(messages.as_mut(), next_notify_in, &mut outputs)
+                .await;
+        }
+    }
+
+    fn process_outputs(
+        &self,
+        outputs: &mut Vec<message_batcher::Output<OwnedCollectdMessage>>,
+    ) -> std::time::Duration {
+        // sentinel value to avoid having to deal with optional timeouts.
+        let mut next_notification_at = self.clock.now() + chrono::Duration::hours(24);
+
+        // Handle outputs
+        for output in outputs.drain(..) {
+            match output {
+                message_batcher::Output::MessageBatch(batch) => {
+                    // Send the current batch to the batch processor
+                    let _ = group_messages(batch)
+                        .map_err(|err| error!("Error while grouping the message batch: {}", err))
+                        .and_then(|group| {
+                            self.sender.send(group).map_err(|err| {
                                 error!("Error while publishing a message batch: {}", err)
-                            });
-                        }
-                        Err(err) => {
-                            error!("Error while building a message batch: {}", err);
-                        }
-                    }
+                            })
+                        });
                 }
-                None => {
-                    //If the message batching loop returns, it means the MQTT connection has closed
-                    error!("MQTT connection closed. Retrying...");
+                message_batcher::Output::NotifyAt(at) => {
+                    next_notification_at = std::cmp::min(next_notification_at, at);
                 }
```

Comment:
At first read, it's not easy to how the value of  `next_notification_at` is used to trigger a timeout in the *input* loop!